File: Model_Improved.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import TimeSeriesSplit
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import lightgbm as lgb
import xgboost as xgb
import statsmodels.api as sm


class RedemptionSalesImprovedModel:
    def __init__(self, X: pd.DataFrame, target_col: str):
        self.X = X.sort_index()
        self.target_col = target_col
        self.results = {}
        self.preds = {}
        self.models = {}
        self.feature_importances_ = {}

    @staticmethod
    def _metrics(y_true, y_pred):
        mask = y_true != 0
        mape = np.mean(np.abs((y_true[mask] - y_pred[mask]) / y_true[mask])) * 100
        return {
            "MSE": mean_squared_error(y_true, y_pred),
            "RMSE": np.sqrt(mean_squared_error(y_true, y_pred)),
            "MAE": mean_absolute_error(y_true, y_pred),
            "MAPE": mape,
            "R2": r2_score(y_true, y_pred),
        }

    def _log(self, model, split, y_true, y_pred, fitted=None):
        self.results.setdefault(model, {})[split] = self._metrics(y_true, y_pred)
        self.preds.setdefault(model, {})[split] = y_pred
        if fitted is not None:
            self.models.setdefault(model, {})[split] = fitted

    def _base(self, y_train, y_test, split):
        stl = sm.tsa.seasonal_decompose(y_train, period=365)
        seas = pd.Series(stl.seasonal, index=y_train.index).clip(lower=0)
        doy_means = seas.groupby(seas.index.dayofyear).mean()
        y_hat = pd.Series(doy_means.reindex(y_test.index.dayofyear).values, index=y_test.index)
        self._log("Base", split, y_test, y_hat)

    def _rf(self, X_train, X_test, y_train, y_test, split):
        model = RandomForestRegressor(n_estimators=200, min_samples_split=20, random_state=42, n_jobs=-1)
        model.fit(X_train, y_train)
        y_hat = pd.Series(model.predict(X_test), index=y_test.index)
        self._log("RandomForest", split, y_test, y_hat, model)
        self.feature_importances_["RandomForest"] = model.feature_importances_

    def _lgbm(self, X_train, X_test, y_train, y_test, split):
        model = lgb.LGBMRegressor(n_estimators=600, learning_rate=0.05, subsample=0.6, max_depth=70,
                                  num_leaves=64, random_state=42)
        model.fit(X_train, y_train)
        y_hat = pd.Series(model.predict(X_test), index=y_test.index)
        self._log("LightGBM", split, y_test, y_hat, model)
        self.feature_importances_["LightGBM"] = model.feature_importances_

    def _xgb(self, X_train, X_test, y_train, y_test, split):
        model = xgb.XGBRegressor(n_estimators=600, learning_rate=0.05, max_depth=6, subsample=0.8,
                                 colsample_bytree=0.8, objective="reg:squarederror", random_state=42, n_jobs=-1)
        model.fit(X_train, y_train)
        y_hat = pd.Series(model.predict(X_test), index=y_test.index)
        self._log("XGBoost", split, y_test, y_hat, model)
        self.feature_importances_["XGBoost"] = model.feature_importances_

    def _ensemble(self, y_test, split):
        rf = self.preds["RandomForest"][split]
        lg = self.preds["LightGBM"][split]
        xb = self.preds["XGBoost"][split]
        ens = (rf + lg + xb) / 3
        self._log("Ensemble", split, y_test, ens)

    def run(self, n_splits=4, test_size=365):
        tscv = TimeSeriesSplit(n_splits=n_splits, test_size=test_size)

        for split, (train_idx, test_idx) in enumerate(tscv.split(self.X)):
            X_train, X_test = self.X.iloc[train_idx], self.X.iloc[test_idx]
            y_train, y_test = X_train[self.target_col], X_test[self.target_col]
            feats_train = X_train.drop(columns=[self.target_col])
            feats_test = X_test.drop(columns=[self.target_col])

            self._base(y_train, y_test, split)
            self._rf(feats_train, feats_test, y_train, y_test, split)
            self._lgbm(feats_train, feats_test, y_train, y_test, split)
            self._xgb(feats_train, feats_test, y_train, y_test, split)
            self._ensemble(y_test, split)

        self.performance_results = (
            pd.DataFrame({m: pd.DataFrame(d).T.mean() for m, d in self.results.items()})
            .T.sort_values("MAPE")
        )

        return self

    def metrics(self):
        return self.performance_results

    def plot(self):
        for model in ["Base", "RandomForest", "LightGBM", "XGBoost", "Ensemble"]:
            for split in self.preds.get(model, {}):
                preds = self.preds[model][split]
                plt.figure(figsize=(15, 4))
                plt.plot(self.X[self.target_col], '.', ms=2, color="grey", label="Observed")
                plt.plot(preds, lw=2, color="red", label=f"{model} Split {split}")
                plt.title(f"{model} - CV split {split}")
                plt.legend(); plt.tight_layout(); plt.show()

    def importance(self):
        feature_names = self.X.drop(columns=[self.target_col]).columns
        fig, axes = plt.subplots(3, 1, figsize=(10, 12))

        for ax, model, color in zip(
            axes, ["RandomForest", "LightGBM", "XGBoost"],
            ["skyblue", "lightgreen", "salmon"]):

            if model in self.feature_importances_:
                importances = self.feature_importances_[model]
                top_features = pd.Series(importances, index=feature_names).nlargest(10)
                ax.barh(top_features.index, top_features.values, color=color)
                ax.set_title(f"Top 10 Feature Importances - {model}")
                ax.invert_yaxis()

        plt.tight_layout()
        plt.show()

File: Model_improved.R
# R version of the RedemptionSalesImprovedModel using functional programming

# Required libraries
library(randomForest)
library(lightgbm)
library(xgboost)
library(Metrics)
library(ggplot2)
library(dplyr)
library(lubridate)


# Convert Timestamp to Date format
df$Timestamp <- as.Date(df$Timestamp)

# ---- Core functions ----

calculate_metrics <- function(y_true, y_pred) {
  mask <- y_true != 0
  mape <- mean(abs((y_true[mask] - y_pred[mask]) / y_true[mask])) * 100
  list(
    MSE = mse(y_true, y_pred),
    RMSE = rmse(y_true, y_pred),
    MAE = mae(y_true, y_pred),
    MAPE = mape,
    R2 = 1 - sum((y_true - y_pred)^2) / sum((y_true - mean(y_true))^2)
  )
}

run_cv_model <- function(data, target_col, n_splits = 4, test_size = 365) {
  data <- data %>% arrange(Timestamp)
  n <- nrow(data)
  split_size <- floor((n - test_size) / n_splits)
  
  results <- list()
  preds <- list()
  feature_importances <- list()
  test_indices <- list()
  
  for (i in seq_len(n_splits)) {
    train_end <- split_size * i
    test_start <- train_end + 1
    test_end <- test_start + test_size - 1
    if (test_end > n) break
    
    train <- data[1:train_end, ]
    test <- data[test_start:test_end, ]
    
    y_train <- train[[target_col]]
    y_test <- test[[target_col]]
    X_train <- train %>% select(-all_of(target_col), -Timestamp)
    X_test <- test %>% select(-all_of(target_col), -Timestamp)
    
    row_idx <- test$Timestamp
    test_indices[[i]] <- row_idx
    
    # Base Model (Day-of-Year mean)
    doy <- yday(train$Timestamp)
    doy_means <- tapply(y_train, doy, mean, na.rm = TRUE)
    base_pred <- doy_means[yday(row_idx)]
    results[["Base"]][[i]] <- calculate_metrics(y_test, base_pred)
    preds[["Base"]][[i]] <- data.frame(index = row_idx, prediction = base_pred)
    
    # Random Forest
    rf <- randomForest(X_train, y_train, ntree = 200, nodesize = 20)
    rf_pred <- predict(rf, X_test)
    results[["RandomForest"]][[i]] <- calculate_metrics(y_test, rf_pred)
    preds[["RandomForest"]][[i]] <- data.frame(index = row_idx, prediction = rf_pred)
    feature_importances[["RandomForest"]] <- setNames(importance(rf)[, 1], rownames(importance(rf)))
    
    
    # LightGBM
    dtrain_lgb <- lgb.Dataset(data = as.matrix(X_train), label = y_train)
    lgb_model <- lgb.train(list(objective = "regression", learning_rate = 0.05,
                                max_depth = 70, num_leaves = 64, subsample = 0.6),
                           dtrain_lgb, nrounds = 600)
    lgb_pred <- predict(lgb_model, as.matrix(X_test))
    results[["LightGBM"]][[i]] <- calculate_metrics(y_test, lgb_pred)
    preds[["LightGBM"]][[i]] <- data.frame(index = row_idx, prediction = lgb_pred)
    lgb_imp <- lgb.importance(lgb_model)
    feature_importances[["LightGBM"]] <- setNames(lgb_imp$Gain, lgb_imp$Feature)
    
    # XGBoost
    dtrain_xgb <- xgb.DMatrix(data = as.matrix(X_train), label = y_train)
    xgb_model <- xgb.train(params = list(objective = "reg:squarederror",
                                         learning_rate = 0.05, max_depth = 6,
                                         subsample = 0.8, colsample_bytree = 0.8),
                           data = dtrain_xgb, nrounds = 600)
    xgb_pred <- predict(xgb_model, xgb.DMatrix(as.matrix(X_test)))
    results[["XGBoost"]][[i]] <- calculate_metrics(y_test, xgb_pred)
    preds[["XGBoost"]][[i]] <- data.frame(index = row_idx, prediction = xgb_pred)
    xgb_imp <- xgb.importance(model = xgb_model)
    feature_importances[["XGBoost"]] <- setNames(xgb_imp$Gain, xgb_imp$Feature)
    
    # Ensemble
    ensemble_pred <- (rf_pred + lgb_pred + xgb_pred) / 3
    results[["Ensemble"]][[i]] <- calculate_metrics(y_test, ensemble_pred)
    preds[["Ensemble"]][[i]] <- data.frame(index = row_idx, prediction = ensemble_pred)
  }
  
  list(
    results = results,
    preds = preds,
    importances = feature_importances,
    summary = summarise_results(results),
    test_indices = test_indices
  )
}



summarise_results <- function(results) {
  m1 <- data.frame(lapply(do.call(rbind, results$Base)%>%
                            as.data.frame(),function(col) unlist(col)))%>%
    colMeans()
  m2 <- data.frame(lapply(do.call(rbind, results$RandomForest)%>%
                            as.data.frame(),function(col) unlist(col)))%>%
    colMeans()
  m3 <- data.frame(lapply(do.call(rbind, results$LightGBM)%>%
                            as.data.frame(),function(col) unlist(col)))%>%
    colMeans()
  m4 <- data.frame(lapply(do.call(rbind, results$XGBoost)%>%
                            as.data.frame(),function(col) unlist(col)))%>%
    colMeans()
  m5 <- data.frame(lapply(do.call(rbind, results$Ensemble)%>%
                            as.data.frame(),function(col) unlist(col)))%>%
    colMeans()
  performance_results <- data.frame(Base = m1, RandomForest = m2, LightGBM = m3,
                                    XGBoost = m4, Ensemble = m5)%>%round(2)
  return(performance_results)
}

plot_predictions <- function(data, preds, target_col) {
  
  for (model_name in names(preds)) {
    for (i in 1:length(preds[[model_name]])) {
      pred_df <- preds[[model_name]][[i]]
      dt <- data.frame(
        index = as.Date(pred_df$index),
        observed = data %>% filter(Timestamp %in% pred_df$index) %>% pull(!!sym(target_col)),
        predicted = pred_df$prediction
      )%>%tibble()
      
      dt_long <- tidyr::pivot_longer(dt, cols = c("observed", "predicted"), names_to = "type", values_to = "value")
      p <- ggplot(dt_long, aes(x = index, y = value, color = type)) +
        geom_point(size = 2.5) +
        #geom_line(lwd = 1.5) +
        scale_color_manual(values = c("observed" = "grey", "predicted" = "red")) +
        ggtitle(paste(model_name, "- Split", i)) +
        theme_bw() +
        theme(legend.title = element_blank(),
            legend.key.size =  unit(1.5, 'cm'),
            legend.text = element_text(size=12),
            axis.title = element_text(size = 15, face = 'bold'),
            axis.text = element_text(size = 13),
            title = element_text(size = 16, face='bold')
            )
        
        print(p)
    }
  }
}



plot_feature_importance <- function(importances) {
  #par(mfrow = c(3, 1))
  for (model in c("RandomForest", "LightGBM", "XGBoost")) {
    if (!is.null(importances[[model]])) {
      imp <- sort(importances[[model]], decreasing = F)[1:10]
      imp_df <- data.frame(Feature = names(imp), Importance = imp)

      #barplot(imp, horiz = TRUE, main = paste("Top 10 -", model), col = "skyblue")
      #barplot(imp, names.arg = names(imp), horiz = TRUE, las = 1,
       #       main = paste("Top 10 -", model), col = "skyblue")

        p <- ggplot(imp_df, aes(x = reorder(Feature, Importance), y = Importance)) +
        geom_bar(stat = "identity", fill = "skyblue") +
        coord_flip() +
        labs(title = paste("Top 10 -", model), x = "Feature", y = "Importance") +
        theme_bw()+
        theme(
            axis.title = element_text(size = 15, face = 'bold'),
            axis.text = element_text(size = 13),
            title = element_text(size = 16, face='bold')
        )
        print(p)

    
    }
  }
  #par(mfrow = c(1, 1))
}